### ISSUE TITLE
Why is my composer.lock file not locking versions down?

### ISSUE THREAD

**Issue by b-hayes** [composer/composer#10457]
My `composer.json`:

```json
{
    "name": "<removed>",
    "description": "<removed>",
    "minimum-stability": "stable",
    "license": "proprietary",
    "autoload": {
        "psr-4": {
            "<removed>": "<removed>"
        }
    },
    "autoload-dev": {
        "psr-4": {
            "<removed>": "tests"
        }
    },
    "scripts": {
        "post-autoload-dump":"git config core.hooksPath .hooks",
        "api:sync": "docker exec -w /app giving_php ./bin/sync-api",
        "docker:start": "docker-compose -p checkin -f docker/docker-compose.yml -f docker/docker-compose.dev.yml up -d",
        "docker:stop": "docker-compose -p checkin -f docker/docker-compose.yml -f docker/docker-compose.dev.yml down",
        "docker:rebuild": "docker-compose -p checkin -f docker/docker-compose.yml -f docker/docker-compose.dev.yml build --force-rm && docker-compose -p checkin -f docker/docker-compose.yml -f docker/docker-compose.dev.yml up -d --remove-orphans",
        "db:migrate": "docker exec -w /app checkin_php ./bin/migrate migrate",
        "db:migrate-all": "composer db:migrate catalog && composer db:migrate tenant",
        "db:rollback": "docker exec -w /app checkin_php ./bin/migrate rollback",
        "db:reset": "docker exec checkin_mysql sh /docker-entrypoint-initdb.d/init.sh",
        "hook:post-commit" : "vendor/bin/common-cs-fixer --container=checkin_php --scan=last --autoFix=true --hookErrors=true",
        "lint:scan-all" : "vendor/bin/common-cs-fixer --container=checkin_php --scan=all",
        "lint:fix-all" : "vendor/bin/common-cs-fixer --container=checkin_php --scan=all --autoFix=true",
        "test": "docker exec -w /app checkin_php ./vendor/bin/phpunit",
        "test:report": "docker exec -w /app checkin_php ./vendor/bin/phpunit --coverage-html test-coverage",
        "build-api": "docker run --rm -v `pwd`:`pwd` -w `pwd` jeanberu/swagger-cli swagger-cli bundle -r -t yaml -o config/api/2021-09-01.yml config/api/src/api.yml",
        "validate-api": "docker run --rm -v `pwd`:`pwd` -w `pwd` jeanberu/swagger-cli swagger-cli validate config/api/src/api.yml"
    },
    "require": {
        "php": "^7.4",
        "ext-json": "*",
        "ext-memcached": "*",
        "ext-pcntl": "*",
        "ext-pdo": "*",
        "tithely/common": "^3.5",
        "tithely/exo": "2.0.*",
        "tithely/picomapper": "1.0.*",
        "tithely/product-clients": "^1.1",
        "tithely/still": "^1.3",
        "elvanto/picodb": "^2.0",
        "alcohol/iso4217": "^3.1",
        "cache/memcached-adapter": "^1.0",
        "hkarlstrom/openapi-validation-middleware": "^0.3",
        "league/oauth2-server": "^8.3",
        "league/route": "^5.1",
        "lstrojny/functional-php": "^1.17",
        "monolog/monolog": "^2.3",
        "northwoods/lazy-middleware": "^2",
        "pimple/pimple": "v3.3.*",
        "respect/validation": "^1.1",
        "sentry/sdk": "^3.1",
        "tuupola/cors-middleware": "^1.2",
        "zendframework/zend-diactoros": "^2.2",
        "zendframework/zend-httphandlerrunner": "^1.1",
        "dragonmantank/cron-expression": "^3.1"
    },
    "require-dev":{
        "tithely/codegen": "1.0.*",
        "phpunit/phpunit": "^9"
    },
    "repositories": [
        {
            "type": "vcs",
            "url": "<removed>"
        },
        {
            "type": "vcs",
            "url": "<removed>"
        },
        {
            "type": "vcs",
            "url": "<removed>"
        },
        {
            "type": "vcs",
            "url": "<removed>"
        }
    ],
    "config": {
        "discard-changes": true,
        "platform": {
            "php": "7.4",
            "ext-memcached": "7.4"
        }
    }
}

```

Output of `composer diagnose`:

```
Checking composer.json: OK
Checking platform settings: OK
Checking git settings: OK
Checking http connectivity to packagist: OK
Checking https connectivity to packagist: OK
Checking github.com oauth access: OK
Checking disk free space: OK
Checking pubkeys: 
Tags Public Key Fingerprint: <removed>
Dev Public Key Fingerprint: <removed>
OK
Checking composer version: OK
Composer version: 2.2.4
PHP version: 7.4 - Package overridden via config.platform, actual: 7.4.3
PHP binary path: /usr/bin/php7.4
OpenSSL version: OpenSSL 1.1.1f  31 Mar 2020
cURL version: 7.68.0 libz 1.2.11 ssl OpenSSL/1.1.1f
zip: extension present, unzip present, 7-Zip not available
```

When I run this command: 

```
composer update
```

I get the following output: 

```
My composer.lock file ends up with varying version constraints that do not match the composer.json for examle:
            "require": {
                "php": "^5.5 || ^7.0 || ^8.0",
                "psr/cache": "^1.0 || ^2.0"
            },...
            "require": {
                "php": "^5.3.2 || ^7.0 || ^8.0"
            },
            "require-dev": {
                "phpstan/phpstan": "^1",
                "phpstan/phpstan-strict-rules": "^1.1",
                "symfony/phpunit-bridge": "^4.2 || ^5"
            },


I then end up with GitHub workflows not being able to resolve dependencies:
(from github)
Run composer install --prefer-dist --no-progress
Installing dependencies from lock file (including require-dev)
Verifying lock file contents can be installed on current platform.
Your lock file does not contain a compatible set of packages. Please run composer update.
Error: Your lock file does not contain a compatible set of packages. Please run composer update.

If I try running composer update again after this nothing changes 🤷‍♂️
```

And I expected this to happen:
I expect the lock file to lock the versions down so that what I install locally is what installs on Github.

Can some please help?

**Comment by stof** [comment]
The dependency resolution is always done for a particular PHP version, not for a range of PHP versions. By default, it uses the version from the PHP runtime. But this can be overridden thanks to the `platform` config option.
What you see in the lock file are the PHP requirements of various packages being installed (and all the examples you gave are indeed compatible with PHP 7.4, which is your target PHP version according to the diagnose output, so I don't understand what you complain about).

**Comment by b-hayes** [comment]
Right then I wonder why the composer install on another system says the dependancies can not be resolved.

**Comment by stof** [comment]
Well, without the error message, I cannot help.
But if you locked your dependency resolved on PHP 7.4 and you try to install them on a different PHP version, they might indeed not be compatible.

**Comment by b-hayes** [comment]
Deleting the lock file before doing composer update fixes the problem.
